L58-python-async-api.md

```diff
@@ -0,0 +1,1739 @@
+        """Describes the length of allowed time remaining for the RPC.
+
+        Returns:
+          A nonnegative float indicating the length of allowed time in seconds
+          remaining for the RPC to complete before it is considered to have
+          timed out, or None if no deadline was specified for the RPC.
+        """
+
+    def cancel(self) -> None:
+        """Cancels the RPC.
+
+        Idempotent and has no effect if the RPC has already terminated.
+        """
+
+    def add_callback(self, callback: Callable[None, None]) -> None:
+        """Registers a callback to be called on RPC termination.
+
+        Args:
+          callback: A no-parameter callable to be called on RPC termination.
+
+        Returns:
+          True if the callback was added and will be called later; False if
+            the callback was not added and will not be called (because the RPC
+            already terminated or some other reason).
+        """
+
+    async def initial_metadata(self) -> Sequence[Tuple[Text, Text]]:
+        """Accesses the initial metadata sent by the server.
+
+        Coroutine continues once the value is available.
+
+        Returns:
+          The initial :term:`metadata`.
+        """
+
+    async def trailing_metadata(self) -> Sequence[Tuple[Text, Text]]:
+        """Accesses the trailing metadata sent by the server.
+
+        Coroutine continues once the value is available.
+
+        Returns:
+          The trailing :term:`metadata`.
+        """
+
+    async def code(self) -> grpc.StatusCode:
+        """Accesses the status code sent by the server.
+
+        Coroutine continues once the value is available.
+
+        Returns:
+          The StatusCode value for the RPC.
+        """
+
+    async def details(self) -> Text:
+        """Accesses the details sent by the server.
+
+        Coroutine continues once the value is available.
+
+        Returns:
+          The details string of the RPC.
+        """
+
+    def __aiter__(self) -> AsyncIterable[Any]:
+        """Returns the async iterable representation that yields messages.
+
+        Returns:
+          An async iterable object that yields messages.
+        """
+
+    async def read(self) -> Any:
+        """Reads one message from the RPC.
+
+        Only one read operation is allowed simultaneously. Mixing new streaming API and old
+        streaming API will resulted in undefined behavior.
+
+        Returns:
+          A response message of the RPC.
+        
+        Raises:
+          An RpcError exception if the read failed.
+        """
+
+    async def write(self, message: Any) -> None:
+        """Writes one message to the RPC.
+
+        Only one write operation is allowed simultaneously. Mixing new streaming API and old
+        streaming API will resulted in undefined behavior.
+
+        Raises:
+          An RpcError exception if the write failed.
+        """
+```
+
+### Server-Side
+
+```Python
+# grpc.aio.Server
+class Server:
```

Comment:
It would be nice if there were some way to get an [`asyncio.Event`](https://docs.python.org/3/library/asyncio-sync.html#asyncio.Event) signalling termination of the server so that I can run another long-running process concurrently with the server in the event loop.

Reply:
Discussed offline. We decided to have `server.wait_for_termination()` return a `Future` instead.

Reply:
As discussed offline, the functionality is similar to `asyncio.Future`, I think the goal you want to achieve can be fulfilled by calling `asyncio.ensure_future` on `server.wait_for_termination`.